numpy/lib/format.py

```diff
@@ -265,13 +265,20 @@ def dtype_to_descr(dtype):
         replicate the input dtype.
 
     """
+    warn_msg = ("metadata on a dtype can be saved but cannot be read. Use "
+               "another form of storage for the array")
     if dtype.names is not None:
         # This is a record array. The .descr is fine.  XXX: parts of the
         # record array with an empty name, like padding bytes, still get
         # fiddled with. This needs to be fixed in the C implementation of
         # dtype().
+        for k in dtype.names:
+            if dtype[k].metadata:
+                warnings.warn (warn_msg, UserWarning, stacklevel=2)
         return dtype.descr
     else:
+        if dtype.metadata:
+            warnings.warn (warn_msg, UserWarning, stacklevel=2)
```

Comment:
Can this ever happen: a dtype without `names` but with `metadata`?

Reply:
Yes, metadata can be attached to any dtype. The main issue is that metadata can be attached to _nested_ dtypes!

Reply:
Yes, metadata can be attached to any dtype. The main issue is that metadata can be attached to nested dtypes!

Reply:
Yes; `dt = np.dtype(int, metadata={})`

Reply:
Subarray dtypes too!

Reply:
Adding tests for these. It turns out `dt = np.dtype(int, metadata={})` will not crash on loading, but it also will not load the metadata. What do you mean by subarray dtypes?